### PR TITLE
Experimental soft label support

### PR DESCRIPTION
**Pull request by furunkel**
Hi Hlib, I have no idea how to run this script, so I blindly put some stuff in. It probably won't run right away, but might contain a few good hints?

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -406,7 +448,7 @@ def numericalize(self, ds):
                 'is_truncated': 'True' if (not (attention_mask == 0).any() ) else 'False', #HACK has to be string so that its ignored by the model
             }
             if not self.no_ground_truth:
-                dct['label'] = labels[ind]
+                dct['labels'] = labels[ind]
```
Comment (furunkel) [inline run.py:451]: Not sure if it was changed to "labels" in a more recent version or if it is changed automatically from 'label' to 'labels' at some point, a bit strange...

File: run.py [hunk 2]
```diff
@@ -107,6 +111,44 @@
 NEXT_FILE_TOKEN = '<nextfile>'
 
 
+class SoftLabelCrossEntropyLoss(nn.Module):
```
Comment (hlibbabii) [inline run.py:114]: What's the difference between `SoftLabelCrossEntropyLoss` and `torch.nn.CrossEntropyLoss`?

**Comment by hlibbabii** [comment]
Hey @furunkel ! Thanks a lot, I'll try running it! Yeah, it's a bit of a mess with running, you set env var `task` e.g. `task=task21`, and to run training you need to replace `evaluation_config` with `training_config`.

Btw, feel free to take full ownership of this code/rewrite everything completely/do not use it at all. Would be still cool though if we could peer-review each other's code, I think it's super useful!

File: run.py (hunk 2 above)
Comment (furunkel) [reply run.py:114]: `nn.CrossEntropyLoss` is optimized for hard labels, i.e., if the target distribution is 1.0 for one class and zero for all others. I guess `nn.BCELoss` would also work; mathematically slightly different, not sure if it matters...

**Comment by hlibbabii** [comment]
Here, it should be 'labels' instead: https://github.com/giganticode/diff-classifier/blob/b1a9aa5aa26df207b4b2029e8f0b32875b426435/run.py#L656

File: run.py (hunk 2 above)
Comment (hlibbabii) [reply run.py:114]: Alright, I see, `nn.BCELoss` uses the sigmoid function, which is better for multilabel classification? 

This is what I had before.

**Comment by hlibbabii** [comment]
Hmm, so I ran 3 models, and compared to `nn.BCELoss`, for 1 model the numbers go up, for 2 others the numbers go down:

models/all_heuristics_without_issues_only_message_soft/eval_results.txt

#before
-eval_loss = 0.5026146769523621
-eval_acc = 0.8606435859422691
-eval_f1 = 0.8486365204314053

#after
+eval_loss = 0.5035837888717651
+eval_acc = 0.8387837562861544
+eval_f1 = 0.8292592658671138

models/only_message_keywords_only_message_soft/eval_results.txt

#before
-eval_loss = 0.534029483795166
-eval_acc = 0.8403534780965669
-eval_f1 = 0.8324021865259706

#after
+eval_loss = 0.5342140197753906
+eval_acc = 0.7889886921891806
+eval_f1 = 0.7828150189123297

models/only_message_keywords_only_message_soft_aug/eval_results.txt

#before
-eval_loss = 0.537394642829895
-eval_acc = 0.7837562861544722
-eval_f1 = 0.7775989341811912

#after
+eval_loss = 0.5377950072288513
+eval_acc = 0.8446266096915788
+eval_f1 = 0.8360687594338152

**Comment by furunkel** [comment]
Hm, interesting. So which one of these runs on BCE vs SoftCE? `nn.BCELoss` wants probabilities, though. Here we feed logits, so we would need `nn.BCEWithLogitsLoss`, if I'm not mistaken...

File: run.py (hunk 2 above)
Comment (furunkel) [reply run.py:114]: BCELoss treats each label as a separate binary variable and averages over them. So, not really a soft label by the books, but maybe close enough...would be interesting to see if it makes a difference...

**Comment by hlibbabii** [comment]
Yeah exactly, I meant `BCEWithLogitsLoss`.  #before is with `BCEWithLogitsLoss` #after is with `SoftCE `
